{"metadata":{"kernelspec":{"language":"python","display_name":"Python 3","name":"python3"},"language_info":{"name":"python","version":"3.7.12","mimetype":"text/x-python","codemirror_mode":{"name":"ipython","version":3},"pygments_lexer":"ipython3","nbconvert_exporter":"python","file_extension":".py"},"kaggle":{"accelerator":"none","dataSources":[{"sourceId":8715559,"sourceType":"datasetVersion","datasetId":5229005},{"sourceId":8720499,"sourceType":"datasetVersion","datasetId":5232712},{"sourceId":8720551,"sourceType":"datasetVersion","datasetId":5232753},{"sourceId":8724638,"sourceType":"datasetVersion","datasetId":5235806}],"dockerImageVersionId":30177,"isInternetEnabled":true,"language":"python","sourceType":"notebook","isGpuEnabled":false}},"nbformat_minor":4,"nbformat":4,"cells":[{"cell_type":"markdown","source":"June 18, 2024 \n\nBy Natalie Ho \n\nAfter working on fast.ai's first module on classifying birds vs. dogs, I decided to see if I could make a module of my own identifying my friends, Koyo and Mitchell! ","metadata":{}},{"cell_type":"code","source":"#NB: Kaggle requires phone verification to use the internet or a GPU. If you haven't done that yet, the cell below will fail\n#    This code is only here to check that your internet is enabled. It doesn't do anything else.\n#    Here's a help thread on getting your phone number verified: https://www.kaggle.com/product-feedback/135367\n\nimport socket,warnings\ntry:\n    socket.setdefaulttimeout(1)\n    socket.socket(socket.AF_INET, socket.SOCK_STREAM).connect(('1.1.1.1', 53))\nexcept socket.error as ex: raise Exception(\"STOP: No internet. Click '>|' in top right and set 'Internet' switch to on\")","metadata":{"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"# It's a good idea to ensure you're running the latest version of any libraries you need.\n# `!pip install -Uqq <libraries>` upgrades to the latest version of <libraries>\n# NB: You can safely ignore any warnings or errors pip spits out about running as root or incompatibilities\nimport os\niskaggle = os.environ.get('KAGGLE_KERNEL_RUN_TYPE', '')\n\nif iskaggle:\n    !pip install -U duckduckgo_search\n    !pip install fastai\n    !pip install fastbook","metadata":{"_kg_hide-input":true,"_kg_hide-output":true,"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"## Step 1: Download images of Koyo","metadata":{}},{"cell_type":"code","source":"from fastbook import *\nfrom fastai.vision.widgets import *\nfrom fastai.vision.all import *","metadata":{"_kg_hide-input":true,"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"'''\nfrom fastdownload import download_url\ndest = 'bird.jpg'\ndownload_url(urls[0], dest, show_progress=False)\n\n\nim = Image.open(dest)\nim.to_thumb(256,256)\n'''","metadata":{"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"Opening a picture of Koyo using Pillow's Image class. ","metadata":{}},{"cell_type":"code","source":"import pandas as pd\nfrom PIL import Image\n\n# Example of opening an image file\nimage_file_path = '/kaggle/input/new-images/Images/Koyo/IMG_4218 Small.jpeg'\nimage = Image.open(image_file_path).to_thumb(256,256)\nimage","metadata":{"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"## Step 2: Train our model","metadata":{}},{"cell_type":"markdown","source":"Used fast.ai's DataBlock to randomly select 80% of images to be the training set and 20% to be the validation set. This function shows a sample of 9 images that classifies if an image contains Mitchell or Koyo in it. \n\n**Challenges**\n* Ran into an UnidentifiedImageError multiple times. I realized that I had to convert some images from HEIC into JPG. But even that wasn't enough! I had to compress all the images to avoid this error. \n* At first I had uploaded two separate folders of images. I realized I had to put the two Koyo and Mitchell subfolders into one larger Images folder as a complete dataset, so that DataBlock could be called and sort images of different categories. ","metadata":{}},{"cell_type":"code","source":"dls = DataBlock(\n    blocks=(ImageBlock, CategoryBlock), \n    get_items=get_image_files, \n    splitter=RandomSplitter(valid_pct=0.2, seed=42),\n    get_y=parent_label,\n    item_tfms=[Resize(192, method='squish')]\n).dataloaders('/kaggle/input/new-images/Images/', bs=40)\n#print(get_items)\ndls.show_batch(max_n=9)\n","metadata":{"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"I thought one of the ways to fix the Unidentified image error was because the image was truncated, but turns out it was a different problem. \n\nHere I am fine tuning my model with resnet18, an image classification set pre-trained on ImageNet dataset. ","metadata":{}},{"cell_type":"code","source":"!pip install pillow\n\nfrom PIL import ImageFile\n\n#ImageFile.LOAD_TRUNCATED_IMAGES = True\n\n\nlearn = vision_learner(dls, resnet18, metrics=error_rate)\nlearn.fine_tune(3)","metadata":{"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"## Step 3: Test the model","metadata":{}},{"cell_type":"code","source":"test = '/kaggle/input/new-images/Images/Koyo/IMG_4594 Small.jpeg'\nis_Koyo,_,probs = learn.predict(PILImage.create(test))","metadata":{"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"print(f\"This is a: {is_Koyo}.\")\nprint(f\"Probability it's a Koyo: {probs[0]:.4f}\")\nimage = Image.open(test).to_thumb(256,256)\nimage","metadata":{"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"## Step 4: Export","metadata":{}},{"cell_type":"code","source":"#export the architecture and trained parameters \nlearn.export() \n#check that this export file exists with the ls method \npath = Path()\npath.ls(file_exts='.pkl')","metadata":{"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"#inference = using a model for getting predictions \nlearn_inf = load_learner(path/'export.pkl')\nlearn_inf.predict(test)","metadata":{"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"categories = {'Koyo', 'Mitchell'}\n\ndef classify_image(img):\n    prediction,index,probability = learn_inf.predict(img)\n    print(prediction,index,probability)\n    result = dict(zip(categories, map(float, probability)))\n    print(result)\n    return result\n\nclassify_image(test)","metadata":{"trusted":true},"execution_count":null,"outputs":[]}]}